src/huggingface_hub/hf_api.py

```diff
@@ -1803,70 +1880,82 @@ def upload_file(
-                    " argument. Be sure to adapt your code!",
-                    FutureWarning,
-                )
-                token, path_or_fileobj, path_in_repo, repo_id = (
-                    path_or_fileobj,
-                    path_in_repo,
-                    repo_id,
-                    token,
-                )
-            else:
-                raise ValueError("Invalid token passed!")
+        commit_summary = (
+            commit_summary
+            if commit_summary is not None
+            else f"Upload {path_in_repo} with huggingface_hub"
+        )
+        operation = CommitOperationAdd(
+            path_or_fileobj=path_or_fileobj,
+            path_in_repo=path_in_repo,
+        )
 
-        # Validate path_or_fileobj
-        if isinstance(path_or_fileobj, str):
-            path_or_fileobj = os.path.normpath(os.path.expanduser(path_or_fileobj))
-            if not os.path.isfile(path_or_fileobj):
-                raise ValueError(f"Provided path: '{path_or_fileobj}' is not a file")
-        elif not isinstance(path_or_fileobj, (RawIOBase, BufferedIOBase, bytes)):
-            # ^^ Test from: https://stackoverflow.com/questions/44584829/how-to-determine-if-file-is-opened-in-binary-or-text-mode
-            raise ValueError(
-                "path_or_fileobj must be either an instance of str or BinaryIO. If you"
-                " passed a fileobj, make sure you've opened the file in binary mode."
+        pr_url = self.create_commit(
+            repo_id=repo_id,
+            repo_type=repo_type,
+            operations=[operation],
+            commit_summary=commit_summary,
+            commit_description=commit_description,
+            token=token,
+            revision=revision,
+            create_pr=create_pr,
+        )
+        return (
+            pr_url
+            if pr_url is not None
+            else hf_hub_url(
+                repo_id, path_in_repo, revision=revision, repo_type=repo_type
             )
+        )
 
-        if repo_type in REPO_TYPES_URL_PREFIXES:
-            repo_id = REPO_TYPES_URL_PREFIXES[repo_type] + repo_id
-
-        revision = revision if revision is not None else "main"
-
-        path = f"{self.endpoint}/api/{repo_id}/upload/{revision}/{path_in_repo}"
-
-        headers = {"authorization": f"Bearer {token}"} if token is not None else None
-
-        if isinstance(path_or_fileobj, str):
-            with open(path_or_fileobj, "rb") as bytestream:
-                r = requests.post(path, headers=headers, data=bytestream)
-        else:
-            r = requests.post(path, headers=headers, data=path_or_fileobj)
-
-        try:
-            r.raise_for_status()
-        except HTTPError as err:
-            if identical_ok and err.response.status_code == 409:
-                from .file_download import hf_hub_url
-
-                return hf_hub_url(
-                    repo_id, path_in_repo, revision=revision, repo_type=repo_type
+    def upload_folder(
```

Comment:
cc @lhoestq 

Only additive for now - meaning it does not delete files that are already present on the repo and not locally